clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -168,76 +329,686 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+                        assignmentReadyToReconcile + " will be handled in the next reconciliation loop.";
+            } else {
+                reason = "Target assignment ready to reconcile is equals to the member current assignment.";
+            }
+            log.debug("Ignoring reconciliation attempt. " + reason);
+            return false;
+        }
+
+        markReconciliationInProgress();
+
+        // Partitions to assign (not previously owned)
+        SortedSet<TopicPartition> addedPartitions = new TreeSet<>(TOPIC_PARTITION_COMPARATOR);
+        addedPartitions.addAll(assignedTopicPartition);
+        addedPartitions.removeAll(ownedPartitions);
+
+        // Partitions to revoke
+        SortedSet<TopicPartition> revokedPartitions = new TreeSet<>(TOPIC_PARTITION_COMPARATOR);
+        revokedPartitions.addAll(ownedPartitions);
+        revokedPartitions.removeAll(assignedTopicPartition);
+
+        log.info("Updating assignment with\n" +
+                        "\tAssigned partitions:                       {}\n" +
+                        "\tCurrent owned partitions:                  {}\n" +
+                        "\tAdded partitions (assigned - owned):       {}\n" +
+                        "\tRevoked partitions (owned - assigned):     {}\n",
+                assignedTopicIdPartitions,
+                ownedPartitions,
+                addedPartitions,
+                revokedPartitions
+        );
+
+        CompletableFuture<Void> revocationResult;
+        if (!revokedPartitions.isEmpty()) {
+            revocationResult = revokePartitions(revokedPartitions);
+        } else {
+            revocationResult = CompletableFuture.completedFuture(null);
+            // Reschedule the auto commit starting from now (new assignment received without any
+            // revocation).
+            commitRequestManager.resetAutoCommitTimer();
+        }
+
+        // Future that will complete when the full reconciliation process completes (revocation
+        // and assignment, executed sequentially)
+        CompletableFuture<Void> reconciliationResult =
+                revocationResult.thenCompose(__ -> {
+                    boolean memberHasRejoined = memberEpochOnReconciliationStart != memberEpoch;
+                    if (state == MemberState.RECONCILING && !memberHasRejoined) {
+                        // Apply assignment
+                        CompletableFuture<Void> assignResult = assignPartitions(assignedTopicPartition,
+                                addedPartitions);
+
+                        // Clear topic names cache only for topics that are not in the subscription anymore
+                        for (TopicPartition tp : revokedPartitions) {
+                            if (!subscriptions.subscription().contains(tp.topic())) {
+                                assignedTopicNamesCache.values().remove(tp.topic());
+                            }
+                        }
```

Comment:
Totally, and actually it made me realize it could be further simplified by retaining the assigned. It is included now in the [follow-up PR](https://github.com/apache/kafka/pull/14857) with the other minor fixes.